diff --git a/browsehappy.com/public_html/inc/locale.php b/browsehappy.com/public_html/inc/locale.php
index 816affd73..1859a6765 100644
--- a/browsehappy.com/public_html/inc/locale.php
+++ b/browsehappy.com/public_html/inc/locale.php
@@ -18,7 +18,7 @@ class Browse_Happy_Locale {
 		self::$active_locales[] = 'en_US';
 
 		if ( isset( $_GET['locale'] ) )
-			$get_locale = preg_replace( '/[^A-Z_-]/i', '', $_GET['locale'] );
+			$get_locale = preg_replace( '/[^A-Z_-]/i', '', sanitize_text_field( wp_unslash( $_GET['locale'] ) ) );
 
 		if ( isset( $get_locale ) && $maybe = self::check_variants( $get_locale ) )
 			self::$locale = $maybe;
@@ -44,7 +44,7 @@ class Browse_Happy_Locale {
 		$locales = array();
 		$variant = '';
 
-		$langs = explode( ',', $_SERVER['HTTP_ACCEPT_LANGUAGE'] );
+		$langs = explode( ',', sanitize_text_field( wp_unslash( $_SERVER['HTTP_ACCEPT_LANGUAGE'] ?? '' ) ) );
 
 		foreach ( $langs as $lang ) {
 			$lang   = str_replace( 'q=', '', $lang );
